include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+} // namespace detail
+
+/// A node type with support for non-trivial edge types.
+template<typename Node,
+         typename EdgeLabel = Empty,
+         bool HasParent = true,
+         size_t SmallSize = 2,
+         typename FinalType = std::false_type,
+         size_t ParentSmallSize = 16,
+         bool ParentHasEntryNode = true>
+class NewNode : public detail::NewNodeBaseTCalc<Node,
+                                                EdgeLabel,
+                                                HasParent,
+                                                SmallSize,
+                                                NewNode,
+                                                FinalType,
+                                                ParentSmallSize,
+                                                ParentHasEntryNode>::Result {
+public:
+  static constexpr bool is_new_node = true;
+  static constexpr bool has_parent = HasParent;
+  using TypeCalc = detail::NewNodeBaseTCalc<Node,
+                                            EdgeLabel,
+                                            HasParent,
+                                            SmallSize,
+                                            NewNode,
+                                            FinalType,
+                                            ParentSmallSize,
+                                            ParentHasEntryNode>;
+  using DerivedType = typename TypeCalc::DerivedType;
+  using Base = typename TypeCalc::Result;
+
+  struct EdgeView {
+    DerivedType &Neighbor;
+    EdgeLabel &Label;
+  };
+  struct OwningEdge {
+    DerivedType *Neighbor;
+    std::unique_ptr<EdgeLabel> Label;
+
+    operator EdgeView() {
+      revng_assert(Neighbor && Label);
+      return EdgeView{ *Neighbor, *Label };
+    }
+    operator EdgeView() const {
+      revng_assert(Neighbor && Label);
+      return EdgeView{ *Neighbor, *Label };
+    }
```

Comment:
Are these really necessary/beneficial?
I often find conversion operators implicit and increasing complexity with little benefit.

Reply:
Well, it doesn't change anything, they might as well be explicit, I guess.

I thought it was fine here as it's pretty much the `std::string`-`std::string_view` relation.
Maybe at one point, we'd introduce one function consuming a view, and another one returning an edge reference, this would make the conversion on the user side transparent.

Anyway, I don't mind both styles, so yeah, I'll make them explicit, no problems.

Reply:
Done.